### PR TITLE
Preserve output of flaky test runs

### LINKED ISSUE #10109: Preserve output of flaky tests
**Description**

When we re-run tests due to flakiness, it seems we lose the output file associated. You can see in failed runs that while there's two XML files for the test (one suffixed with -FLAKY), we only have the successful run's output. This makes diagnosing flaky tests difficult unless we're lucky and they are easy to reproduce.

To save us time, we should figure out how to preserve the output of the test between re-runs. Maybe a test listener can do this, or anything which somehow gets notified that a test will run again.

### PR DESCRIPTION
**Pull request by npepinpe**
## Description

This PR numbers the output of all test runs, such that instead of writing to `MyTest-output.txt`, it now writes to `MyTest-1-output.txt`. This will ensure that when we re-run a failing test, the next run will write to `MyTest-2-output.txt`.

In more details:

To ensure that we preserve test output of flaky tests - that is, the output of a test even if it's re-run - we use a custom implementation of the `ConsoleOutputReporter`. This implementation is a simple delegating implementation, except for to return a custom listener which will ensure that all run outputs are numbered (e.g. 1, 2, 3, etc.) for each test. The custom listener is also a delegating one, and simply hooks into the lifecycle of the canonical implementation. After writing the output (e.g. `MyTest-output.txt`), the listener will compute the run count from the file system (to avoid caching an inordinate amount of test files) and move to the file to its proper numbered name, e.g. `MyTest-1-output.txt` or `MyTest-2-output.txt`).

As the canonical implementation is not meant to be extended, we opted for this delegating pattern which is a little hacky (as we "predict" the file name of the report via reflection), but it's a good compromise in terms of keeping up with Surefire improvements and maintaining our complete implementation.

NOTE: since the implementation moves the file after it's originally written to `MyTest-output.txt`, there is a race condition where the last run may still be un-numbered, but I think this is acceptable, and we're still fixing the problem.

## Related issues

closes #10109 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

**Comment by npepinpe** [comment]
To test, you can use the following:

Replace the following file:

<details><summary>util/src/test/java/io/camunda/zeebe/util/buffer/BufferUtilTest.java</summary>

```java
/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
 * one or more contributor license agreements. See the NOTICE file distributed
 * with this work for additional information regarding copyright ownership.
 * Licensed under the Zeebe Community License 1.1. You may not use this file
 * except in compliance with the Zeebe Community License 1.1.
 */
package io.camunda.zeebe.util.buffer;

import static io.camunda.zeebe.util.StringUtil.getBytes;
import static io.camunda.zeebe.util.buffer.BufferUtil.cloneBuffer;
import static org.assertj.core.api.Assertions.assertThat;

import java.util.concurrent.atomic.AtomicInteger;
import org.agrona.DirectBuffer;
import org.agrona.ExpandableArrayBuffer;
import org.agrona.ExpandableDirectByteBuffer;
import org.agrona.MutableDirectBuffer;
import org.agrona.concurrent.UnsafeBuffer;
import org.junit.Test;

public final class BufferUtilTest {
  protected static final byte[] BYTES1 = getBytes("foo");
  protected static final byte[] BYTES2 = getBytes("bar");
  protected static final byte[] BYTES3 = new byte[BYTES1.length + BYTES2.length];
  private static final AtomicInteger TOGGLE = new AtomicInteger();

  static {
    System.arraycopy(BYTES1, 0, BYTES3, 0, BYTES1.length);
    System.arraycopy(BYTES2, 0, BYTES3, BYTES1.length, BYTES2.length);
  }

  @Test
  public void testEquals() {
    assertThat(BufferUtil.contentsEqual(asBuffer(BYTES1), asBuffer(BYTES1))).isTrue();
    assertThat(BufferUtil.contentsEqual(asBuffer(BYTES1), asBuffer(BYTES2))).isFalse();
    assertThat(BufferUtil.contentsEqual(asBuffer(BYTES1), asBuffer(BYTES3))).isFalse();
    assertThat(BufferUtil.contentsEqual(asBuffer(BYTES3), asBuffer(BYTES1))).isFalse();
  }

  @Test
  public void testCloneUnsafeBuffer() {
    // given
    final DirectBuffer src = new UnsafeBuffer(BYTES1);

    // when
    final DirectBuffer dst = cloneBuffer(src);

    // then
    assertThat(dst).isNotSameAs(src).isEqualTo(src).hasSameClassAs(src);
  }

  @Test
  public void testCloneExpandableArrayBuffer() {
    // given
    final MutableDirectBuffer src = new ExpandableArrayBuffer(BYTES1.length);
    src.putBytes(0, BYTES1);

    // when
    final DirectBuffer dst = cloneBuffer(src);

    // then
    assertThat(dst).isNotSameAs(src).isEqualTo(src).hasSameClassAs(src);
  }

  @Test
  public void shouldReturnCopyOfByteArrayWhenWrappingPartialArray() {
    // given
    final byte[] expected = new byte[BYTES1.length - 1];
    final DirectBuffer buffer = new UnsafeBuffer();
    buffer.wrap(BYTES1, 1, BYTES1.length - 1);

    // when
    final byte[] bytes = BufferUtil.bufferAsArray(buffer);
    System.arraycopy(BYTES1, 1, expected, 0, expected.length);

    // then
    assertThat(buffer.byteArray()).isEqualTo(BYTES1);
    assertThat(bytes).isEqualTo(expected);
    assertThat(bytes).isNotEqualTo(BYTES1);
  }

  @Test
  public void shouldReturnNewByteArrayWhenNotWrappingByteArray() {
    // given
    final MutableDirectBuffer buffer = new ExpandableDirectByteBuffer();
    buffer.putBytes(0, BYTES1);

    // when
    final byte[] bytes = BufferUtil.bufferAsArray(buffer);

    // then
    assertThat(buffer.byteArray()).isNull();
    for (int i = 0; i < BYTES1.length; i++) {
      assertThat(bytes[i]).isEqualTo(buffer.getByte(i));
    }
  }

  /**
   * Returning the wrapped array is an inconsistency in the API, because we cannot safely modify the
   * returned array.
   */
  @Test
  public void shouldNotReturnWrappedArrayIfBufferOffsetIsZero() {
    // given
    final DirectBuffer buffer = new UnsafeBuffer(BYTES1);

    // when
    final byte[] bytes = BufferUtil.bufferAsArray(buffer);

    // then
    assertThat(buffer.byteArray()).isEqualTo(BYTES1);
    assertThat(bytes).isNotSameAs(BYTES1);
  }

  @Test
  public void shouldBeFlakyAndPreserveOutput() {
    final int val = TOGGLE.getAndIncrement();
    System.out.println("Run: " + val);
    if (val < 4) {
      assertThat(true).isFalse();
    }

    System.out.println("Success");
    assertThat(true).isTrue();
  }

  public DirectBuffer asBuffer(final byte[] bytes) {
    return new UnsafeBuffer(bytes);
  }
}
```

</details>

Then make sure to clean the util test reports to make it easier to evaluate the whole thing:

```shell
rm -rf util/target/surefire-reports
```

Then run the tests (building the `build-tools` as well):

```shell
mvn install -am -pl util/ -Dtest=BufferUtilTest -Dsurefire.rerunFailingTestsCount=5 -Dsurefire.redirectOutputToFile=true -DfailIfNoTests=false
```

Then check the `util/target/surefire-report` and you will see something like 

```
❯ ls util/target/surefire-reports                                                                                                               
io.camunda.zeebe.util.buffer.BufferUtilTest-1-output.txt  io.camunda.zeebe.util.buffer.BufferUtilTest-3-output.txt  io.camunda.zeebe.util.buffer.BufferUtilTest-5-output.txt  TEST-io.camunda.zeebe.util.buffer.BufferUtilTest.xml
io.camunda.zeebe.util.buffer.BufferUtilTest-2-output.txt  io.camunda.zeebe.util.buffer.BufferUtilTest-4-output.txt  io.camunda.zeebe.util.buffer.BufferUtilTest.txt
```

**Comment by npepinpe** [comment]
I'm not sure why revapi breaks a bunch of stuff now. I didn't check anything...

**Comment by npepinpe** [comment]
I guess the new dependencies must be adding some transitive dependencies

**Comment by Zelldon** [comment]
I would prefer to just not rerun flaky tests

**Comment by npepinpe** [comment]
Happy to have this discussion again, we have enough new team members

**Comment by npepinpe** [comment]
Anyway, I would still say let's merge this as there is immediate value to it for all our flaky tests. If you can get consensus from the team fast enough on simply not re-running tests, then sure, we can close this and go for that.

**Comment by megglos** [comment]
> Anyway, I would still say let's merge this as there is immediate value to it for all our flaky tests. If you can get consensus from the team fast enough on simply not re-running tests, then sure, we can close this and go for that.

agree with Nicolas here, as this change is ready and brings immediate benefit I'd like to get it in. Decoupled from the discussion on how to best handle flaky tests going forward.

**Review by megglos — APPROVED** [review]
LGTM! Do we have a sample run with flaky tests :D maybe we can provoke some with higher fork count? but I'm also fine with merging first and provoking them after in a test PR.

File: build-tools/src/main/java/io/camunda/zeebe/ZeebeConsoleOutputFileReporter.java [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+    final Path directory = path.getParent();
+    int latestRunCount = 0;
+
+    try (final DirectoryStream<Path> files =
+        Files.newDirectoryStream(directory, p -> filterTestOutputFiles(fileName, p))) {
+      for (final Path file : files) {
+        final Matcher matcher = RUN_COUNT_MATCHER.matcher(file.getFileName().toString());
+        if (matcher.find()) {
+          final String runCount = matcher.group(1);
+          latestRunCount = Math.max(latestRunCount, Integer.parseInt(runCount));
```
[the remark this answers is not in the document]
Comment (npepinpe) [reply build-tools/src/main/java/io/camunda/zeebe/ZeebeConsoleOutputFileReporter.java:135]: It's actually caught indirectly, foolish CodeQL :neutral_face:

**Comment by megglos** [comment]
tested it by provoking flakes due to higher forks 
https://github.com/camunda/zeebe/actions/runs/3142212600 looks good!

**Comment by npepinpe** [comment]
I did not test it with a high fork count true, so it might be interesting to do this.

File: parent/pom.xml [hunk 2]
```diff
@@ -1216,8 +1216,16 @@
               <usePhrasedTestCaseClassName>true</usePhrasedTestCaseClassName>
               <usePhrasedTestCaseMethodName>true</usePhrasedTestCaseMethodName>
             </statelessTestsetReporter>
+            <consoleOutputReporter implementation="io.camunda.zeebe.ZeebeConsoleOutputReporter"></consoleOutputReporter>
```
Comment (megglos) [inline parent/pom.xml:1219]: just noticed by this failure https://github.com/camunda/zeebe/actions/runs/3142205966/jobs/5105531759 that currently the unit test job doesn't use reruns and flaky test detection. Might be worth reenabling right now for consistency.

Comment (npepinpe) [reply parent/pom.xml:1219]: Strange, as we had that in Jenkins :thinking: I would've thought we would keep it as well for GHA

Comment (megglos) [reply parent/pom.xml:1219]: I suspect this got lost with the recent amount of changes, in this case the merge of the split unit tests jobs into one.

Comment (oleschoenburg) [reply parent/pom.xml:1219]: I think I never enabled it for unit tests, even in the original GHA workflows :see_no_evil:

Comment (npepinpe) [reply parent/pom.xml:1219]: I enabled it for unit, client, and smoke tests.

File: .github/workflows/test.yml [hunk 3]
```diff
@@ -145,7 +145,7 @@ jobs:
       - name: Maven Test Build
         run: >
           mvn -T2 -B --no-snapshot-updates
-          -D skipChecks
+          -D skipChecks -D surefire.rerunFailingTestsCount=3
```
Comment (megglos) [inline .github/workflows/test.yml:148]: we also need to enable the profile `extract-flaky-tests` then?

Comment (npepinpe) [reply .github/workflows/test.yml:148]: Correct - for some reason the client tests were already doing that :see_no_evil:

**Comment by npepinpe** [comment]
Is this fine to merge, or...?

**Comment by npepinpe** [comment]
bors merge

**Pull request merged**